clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,758 @@
+    return 3;
+  default:
+    llvm_unreachable("unexpected SYCLBinaryImageFormat");
+  }
+}
+
+// Wrapper helper class that creates all LLVM IRs wrapping given images.
+// Note: All created structures, "_pi_device_*", "__sycl_*" and "__tgt*" names
+// in this implementation are aligned with "sycl/include/sycl/detail/pi.h". If
+// you want to change anything then you MUST coordinate changes with SYCL
+// Runtime ABI.
+struct Wrapper {
+  Module &M;
+  LLVMContext &C;
+  SYCLWrappingOptions Options;
+
+  Wrapper(Module &M, const SYCLWrappingOptions &Options)
+      : M(M), C(M.getContext()), Options(Options) {}
+
+  // struct _pi_device_binary_property_struct {
+  //   char *Name;
+  //   void *ValAddr;
+  //   uint32_t Type;
+  //   uint64_t ValSize;
+  // };
+  StructType *getSyclPropTy() {
+    return StructType::create({Type::getInt8PtrTy(C), Type::getInt8PtrTy(C),
+                               Type::getInt32Ty(C), Type::getInt64Ty(C)},
+                              "_pi_device_binary_property_struct");
+  }
+
+  PointerType *getSyclPropPtrTy() {
+    return PointerType::getUnqual(getSyclPropTy());
+  }
+
+  // struct _pi_device_binary_property_set_struct {
+  //   char *Name;
+  //   _pi_device_binary_property_struct* PropertiesBegin;
+  //   _pi_device_binary_property_struct* PropertiesEnd;
+  // };
+  StructType *getSyclPropSetTy() {
+    return StructType::create(
+        {Type::getInt8PtrTy(C), getSyclPropPtrTy(), getSyclPropPtrTy()},
+        "_pi_device_binary_property_set_struct");
+  }
+
+  PointerType *getSyclPropSetPtrTy() {
+    return PointerType::getUnqual(getSyclPropSetTy());
+  }
+
+  IntegerType *getSizeTTy() {
+    switch (M.getDataLayout().getPointerTypeSize(Type::getInt8PtrTy(C))) {
+    case 4:
+      return Type::getInt32Ty(C);
+    case 8:
+      return Type::getInt64Ty(C);
+    }
+    llvm_unreachable("unsupported pointer type size");
+  }
+
+  SmallVector<Constant *, 2> getSizetConstPair(size_t First, size_t Second) {
+    IntegerType *SizeTTy = getSizeTTy();
+    return SmallVector<Constant *, 2>{ConstantInt::get(SizeTTy, First),
+                                      ConstantInt::get(SizeTTy, Second)};
+  }
+
+  // struct __tgt_offload_entry {
+  //   void *addr;
+  //   char *name;
+  //   size_t size;
+  //   int32_t flags;
+  //   int32_t reserved;
+  // };
+  StructType *getEntryTy() {
+    return StructType::create("__tgt_offload_entry", Type::getInt8PtrTy(C),
+                              Type::getInt8PtrTy(C), getSizeTTy(),
+                              Type::getInt32Ty(C), Type::getInt32Ty(C));
+  }
+
+  PointerType *getEntryPtrTy() { return PointerType::getUnqual(getEntryTy()); }
+
+  // SYCL specific image descriptor type.
+  // struct __tgt_device_image {
+  //   /// version of this structure - for backward compatibility;
+  //   /// all modifications which change order/type/offsets of existing fields
+  //   /// should increment the version.
+  //   uint16_t Version;
```

Comment:
> We should investigate if we really have a case where device images within the same binary have different versions and see how can we align with the community.

Consider the case where a AOT compiled MKL library with an old compiler is linked into with a newer compiler.

Reply:
> > We should investigate if we really have a case where device images within the same binary have different versions and see how can we align with the community.
> 
> Consider the case where a AOT compiled MKL library with an old compiler is linked into with a newer compiler.

I'm still not sure that it is the use case for the field. Let's review all possible combinations:

#### We link "old" object file using new offloading model

Object files contain device code, wrapping is done at link stage, so there will be a single version set for all device images produced, because they are all produced by the new offloading model regardless of their source.

#### We link with a shared library built using old offloading model

In that case we don't link device code from the app to the library, still no use for version.

#### We link with an early-AOT compiled object file/archive that was produced by old version of the toolchain

Device code in that object file/archive is already wrapped, I assume and it has version already encoded in it by the old version of the toolchain.

--- 

As you can see, we just hardcode some version in the final binary regardless of input device images:

https://github.com/intel/llvm/blob/6641be012aed1b1fea84be1c64b94469ecc7f720/clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp#L547-549

https://github.com/intel/llvm/blob/6641be012aed1b1fea84be1c64b94469ecc7f720/clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp#L620

My guess it that the idea was to allow modifications/evolution of `__tgt_device_image` and `__tgt_bin_desc` data structures by introducing the version into it. I.e. runtime could have read the version first and then re-interpret the data structure differently. But it still looks a bit weird to me: layout of the struct can't be changed so that version field moves from revision to revision. RT has to be really careful *not* to dereference anything besides version before it understand which exact layout is there.

I'm not sure what is the upstream policy for breaking changes, but we can't allow breaking changes on our side (unless otherwise specified). This is a topic which should probably be brought up with the community to understand their stand on the problem and agree on a common solution.

I don't quite like the fact that new version means new struct layout. We could try and use approach similar to [Level Zero](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-context-desc-t):
```
struct ze_context_desc_t {
  ze_structure_type sType;
  const void *pNext;
  ze_context_flags flags;
};
```

Data structure is a set of small chained descriptors where each descriptor is guaranteed to contain `sType` and `pNext` fields at the beginning. When `ze_context_desc_t` is passed, runtime reads its `reinterpret_cast<const ze_context_desc_t *>(pNext)->sType` (with `nullptr` check first of course) to understand what are the extensions to that struct to parse them.

This way, the layout is always the same, old runtime can just ignore unknown structure types, new runtime can on-the-fly generate missing pieces (if the app was produced by an old toolchain). There is no need for version field, layout is stable. The downside is more complicated data structure which may be way more scattered in memory, plus both the compiler and runtime should agree on `ze_structure_type` and each particular extension structs. Interface is bigger, but at the same time it is easier to handle when it comes to version mismatch.

Reply:
Is there much sense to drop extra `Version` since we add no new functionality and add some burden to support both variants?

Reply:
The structure will look a bit nicer and RT handling will become a bit ugly.

Reply:
We could introduce a temporary command line option `-pi-version=` with values `1.2` and `2.0`. When RT adds support for `2.0` then we might remove this option. `-pi-version=1.2` will allow us to test apps during migration.